Add playtime command

melluh opened the pull request:
<!--

EssentialsX feature submission guide
====================================

NOTE: Failure to fill out this template properly may result in your PR being
      delayed or ignored without warning.

NOTE: Don't type between any arrows in the template, as this text will be
      hidden. This includes this header block and any other explanation text
      blocks.

Want to discuss your PR before submitting it? Join the EssentialsX Development
server: https://discord.gg/CUN7qVb


EssentialsX is GPL
------------------

By contributing to EssentialsX, you agree to license your code under the
GNU General Public License version 3, which can be found at the link below:
https://github.com/EssentialsX/Essentials/blob/2.x/LICENSE


Instructions
------------

If you are submitting a new feature, please follow the following steps:

1.  Fill out the template in full.
      This includes providing screenshots and a link to the original feature 
      request. If there isn't an existing feature request, we strongly
      recommend opening a new feature request BEFORE opening your PR to
      implement it, as this allows us to review whether we're likely to accept
      your feature in advance, and also allows us to discuss possible
      implementations for the feature. If there is no associated feature
      request, your PR may be delayed or rejected without warning.
      
      You can open a new feature request by following this link:
      https://github.com/EssentialsX/Essentials/issues/new/choose

2.  If you are fixing a performance issue, please use the "Bug fix" PR template
      instead. The "bug fix" template is better suited to performance issues.

3.  Include a demonstration.
      If you are adding commands, please provide screenshots and/or a video
      demonstration of the feature. Similarly, if you are adding a new API,
      please include a link to example code that takes advantage of your
      proposed API. This will aid us in reviewing PRs and speed up the process
      significantly.

-->

### Information

<!--
    Replace #nnnn with the number of the original issue. If this PR implements
    features from multiple issues, you should repeat the phrase "closes #nnnn"
    for each issue. 
-->

This PR closes #3827. 

### Details

**Proposed feature:**  
<!-- Type a description of your proposed feature below this line. -->
Allow players to view their own or another player's playtime.

**Environments tested:**    

<!-- Type the OS you have used below. -->
OS: Windows 10

<!-- Type the JDK version (from java -version) you have used below. -->
Java version: ``java version "16.0.1" 2021-04-20``

<!--
    Put an "x" inside the boxes for the server software you have tested this 
    bug fix on. If this feature does not apply to a server, strike through the server software using ~~strikethrough~~. If you have tested on other
    environments, add a new line with relevant details.
-->
- [x] Most recent Paper version (1.17.1, git-Paper-fdd5c65)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8


**Demonstration:**  
<!--
    Below this block, include screenshots/log snippets from before and after as
    necessary. If you have created or used a test case plugin, please link to a
    download of the plugin, source code and exact version used where possible.
-->
![2021-10-05_22 11 40](https://user-images.githubusercontent.com/18401198/136095685-7d00e789-c611-4186-bf18-0a04c62636fd.png)
![2021-10-05_22 11 47](https://user-images.githubusercontent.com/18401198/136095691-4b4ae5f6-6d68-44b6-ba51-93468c4018be.png)

melluh committed:
Add playtime command

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java:
```diff
@@ -0,0 +1,54 @@
+import com.earth2me.essentials.utils.EnumUtil;
+
+public class Commandplaytime extends EssentialsCommand {
+
+    // For some reason, in 1.13 PLAY_ONE_MINUTE = ticks played = what used to be PLAY_ONE_TICK
```
```suggestion
    // For some reason, in 1.13 PLAY_ONE_MINUTE = ticks played = what used to be PLAY_ONE_TICK
```

JRoy commented on Essentials/src/main/resources/messages_en.properties:
```diff
@@ -1481,3 +1481,6 @@ years=years
 youAreHealed=§6You have been healed.
 youHaveNewMail=§6You have§c {0} §6messages\! Type §c/mail read§6 to view your mail.
 xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.
+playtimeCommandDescription=Shows a player''s playtime
+playtimeCommandUsage=/<command> [player]
+playtime=§6{0}§6''s Playtime\:§r {1}
```
Do not add messages to this file, add them to `messages.properties`, not `messages_en.properties`.
```suggestion
```

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java:
```diff
@@ -0,0 +1,54 @@
+    
+    @Override
+    protected List<String> getTabCompleteOptions(final Server server, final CommandSource sender, final String commandLabel, final String[] args) {
+        if (args.length == 1) {
```
```suggestion
        if (args.length == 1 && sender.isAuthorized("essentials.playtime.others", ess)) {
```

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java:
```diff
@@ -0,0 +1,54 @@
+    protected void run(Server server, CommandSource sender, String commandLabel, String[] args) throws Exception {
+        final User target;
+        
+        if (args.length >= 1) {
+            target = this.getPlayer(server, sender, args[0]);
+            if (target == null) {
+                throw new PlayerNotFoundException();
+            }
+        } else if (sender.isPlayer()) {
+            target = ess.getUser(sender.getPlayer());
+        } else {
+            throw new NotEnoughArgumentsException();
+        }
```
```suggestion
        if (args.length > 0) {
            target = getPlayer(server, sender, args, 0, false, false);
        } else if (sender.isPlayer()) {
            target = sender.getUser(ess);
        } else {
            throw new NotEnoughArgumentsException();
        }
```

melluh committed:
Apply suggestions from code review

Co-authored-by: Josh Roy <10731363+JRoy@users.noreply.github.com>

melluh committed:
Re-add messages to correct file

JRoy commented on Essentials/src/main/resources/messages.properties:
```diff
@@ -1491,3 +1491,6 @@ years=years
 xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.
+playtimeCommandDescription=Shows a player''s playtime
+playtimeCommandUsage=/<command> [player]
+playtime=\u00a76{0}\u00a76''s Playtime\:\u00a7r {1}
```
please add these in alphabetical order

JRoy commented on Essentials/src/main/resources/messages_en.properties:
```diff
@@ -1480,4 +1480,4 @@ year=year
 youAreHealed=§6You have been healed.
 youHaveNewMail=§6You have§c {0} §6messages\! Type §c/mail read§6 to view your mail.
-xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.
+xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.
```
```suggestion
xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.

```

JRoy committed:
Update Essentials/src/main/resources/messages_en.properties

melluh committed:
Fix issues

- Made it actually compile (oops sorry)
- Handle permissions correctly
- Put new messages in the right place in message.properties

JRoy commented on Essentials/src/main/resources/messages.properties:
```diff
@@ -872,6 +872,9 @@ playerTempBanned=\u00a76Player \u00a7c{0}\u00a76 temporarily banned \u00a7c{1}\u
 playerUnbanned=\u00a76Player\u00a7c {0} \u00a76unbanned\u00a7c {1}
 playerUnmuted=\u00a76You have been unmuted.
+playtimeCommandDescription=Shows a player''s playtime
+playtimeCommandUsage=/<command> [player]
```
```suggestion
playtimeCommandUsage=/<command> [player]
playtimeCommandUsage1=/<command>
playtimeCommandUsage1Description=Shows your time played in game
playtimeCommandUsage2=/<command> <player>
playtimeCommandUsage2Description=Shows the specified player''s time played in game
```

JRoy commented on Essentials/src/main/resources/messages.properties:
```diff
@@ -872,6 +872,9 @@ playerTempBanned=\u00a76Player \u00a7c{0}\u00a76 temporarily banned \u00a7c{1}\u
 playerUnbanIpAddress=\u00a76Player\u00a7c {0} \u00a76unbanned IP\:\u00a7c {1}
 playerUnbanned=\u00a76Player\u00a7c {0} \u00a76unbanned\u00a7c {1}
 playerUnmuted=\u00a76You have been unmuted.
+playtimeCommandDescription=Shows a player''s playtime
```
```suggestion
playtimeCommandDescription=Shows a player''s time played in game
```

JRoy commented on Essentials/src/main/resources/plugin.yml:
```diff
@@ -344,6 +344,10 @@ commands:
     usage: /<command>
     aliases: [echo,eecho,eping,pong,epong]
+  playtime:
+    description: Shows a player's playtime
```
```suggestion
    description: Shows a player's time played in game
```

JRoy commented on Essentials/src/main/resources/messages.properties:
```diff
@@ -872,6 +872,9 @@ playerTempBanned=\u00a76Player \u00a7c{0}\u00a76 temporarily banned \u00a7c{1}\u
 playerUnmuted=\u00a76You have been unmuted.
+playtimeCommandDescription=Shows a player''s playtime
+playtimeCommandUsage=/<command> [player]
+playtime=\u00a76{0}\u00a76''s Playtime\:\u00a7r {1}
```
```suggestion
playtime=\u00a76Playtime\:\u00a7c {0}
playtimeOther=\u00a76Playtime of {1}\u00a76\:\u00a7c {0}
```

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    protected void run(Server server, CommandSource sender, String commandLabel, String[] args) throws Exception {
+        final IUser target;
+        
+        if (args.length > 0 && sender.isAuthorized("essentials.playtime.others", ess)) {
+            target = getPlayer(server, sender, args, 0);
+        } else if (sender.isPlayer()) {
+            target = sender.getUser(ess);
+        } else {
+            throw new NotEnoughArgumentsException();
+        }
+
+        final long playtimeMs = System.currentTimeMillis() - (target.getBase().getStatistic(PLAY_ONE_TICK) * 50);
+        sender.sendMessage(tl("playtime", target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```
```suggestion
        final IUser target;
        final String key;
        
        if (args.length > 0 && sender.isAuthorized("essentials.playtime.others", ess)) {
            target = getPlayer(server, sender, args, 0);
            key = "playtimeOther";
        } else if (sender.isPlayer()) {
            target = sender.getUser(ess);
            key = "playtime";
        } else {
            throw new NotEnoughArgumentsException();
        }

        final long playtimeMs = System.currentTimeMillis() - (target.getBase().getStatistic(PLAY_ONE_TICK) * 50);
        sender.sendMessage(tl("playtime", target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```

JRoy committed:
Apply suggestions from code review

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java:
```diff
@@ -0,0 +1,53 @@
+        }
+
+        final long playtimeMs = System.currentTimeMillis() - (target.getBase().getStatistic(PLAY_ONE_TICK) * 50);
+        sender.sendMessage(tl("playtime", target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```
```suggestion
        sender.sendMessage(tl(key, target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```

JRoy committed:
Update Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java:
```diff
@@ -0,0 +1,53 @@
+        }
+
+        final long playtimeMs = System.currentTimeMillis() - (target.getBase().getStatistic(PLAY_ONE_TICK) * 50);
+        sender.sendMessage(tl(key, target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```
```suggestion
        sender.sendMessage(tl(key, DateUtil.formatDateDiff(playtimeMs), target.getDisplayName()));
```

JRoy committed:
Update Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java

JRoy committed:
Merge branch '2.x' into 2.x

ghost commented:
@MisterMelDev, would it be possible to add offline playtime support? This would be insanely useful.

melluh commented:
> @MisterMelDev, would it be possible to add offline playtime support? This would be insanely useful.

The way I implemented this, it currently wouldn't be possible. I'm using the built-in Bukkit statistics system, which also contains a statistic for playtime, but this can only be retrieved when the player is online. The plugin would have to store the playtime in the player data or something. I'm a bit busy currently, but maybe someone else could make a PR for that?

serious-angel commented:
> The way I implemented this

Greetings! Just in case if still interested in offline data, there are a few implementation like the following wonderful one for a reference:
- <https://www.spigotmc.org/resources/playtime.26016>;
- <https://github.com/Twi5TeD/PlayTime>.
